### ISSUE TITLE
flat_map bug?

### ISSUE THREAD

**Issue by trans** [crystal-lang/crystal#3415]
Is `flat_map` supposed to not handle elements that are not in arrays?

```ruby
a = [ [1,2,3], 4, 5]
x = a.flat_map do |x|
  x
end
p x
```

```
instantiating 'Array(Array(Int32) | Int32)#flat_map()'

x = a.flat_map do |x|
      ^~~~~~~~

in /opt/crystal/src/enumerable.cr:255: instantiating 'each()'

    each { |e| ary.concat(yield e) }
    ^~~~

in /opt/crystal/src/indexable.cr:151: instantiating 'each_index()'

    each_index do |i|
    ^~~~~~~~~~

in /opt/crystal/src/indexable.cr:151: instantiating 'each_index()'

    each_index do |i|
    ^~~~~~~~~~

in /opt/crystal/src/enumerable.cr:255: instantiating 'each()'

    each { |e| ary.concat(yield e) }
    ^~~~

in /opt/crystal/src/enumerable.cr:255: no overload matches 'Array(Int32)#concat' with type (Array(Int32) | Int32)
Overloads are:
 - Array(T)#concat(other : Array)
 - Array(T)#concat(other : Enumerable)
Couldn't find overloads for these types:
 - Array(Int32)#concat(other : Int32)

    each { |e| ary.concat(yield e) }
```

**Comment by asterite** [comment]
Right now `flat_map` expects the block to return an array or an enumerable. Maybe we could change it to check if it's an enumerable or not.

**Comment by trans** [comment]
Perhaps unnecessary. I am going to bring up the topic of Functors and Monads in another issue, and that would no doubt effect this behavior.

**Comment by trans** [comment]
On second thought I won't make another issue. I just sate it here. 

Maybe every object should have a #flat_map method, to indicate how it is to treated by flat_map. So in the case of say an Int32 (and generally all scalar values) it just maps once over itself.

    def flat_map
      yield self
    end

That would avoid special ifs in Enumerable#flat_map and allow objects to define their own behaviors.

**Comment by asterite** [comment]
Fixed by #3427
